test/unit/hyperloglog.c

```diff
@@ -0,0 +1,362 @@
+ * The resulting error is the average error of all cardinalities.
+ */
+void
+measure_dense_hll_estimation_error(int prec, size_t max_card,
+				   size_t n_points, size_t sets_per_point,
+				   struct est_errors *res, FILE *output)
+{
+	double max_err_sum = 0;
+	double std_err_sum = 0;
+	const double card_step = 1.f * max_card / n_points;
+	for (size_t n = 0; n < n_points; ++n) {
+		size_t card = card_step * n;
+		double error[sets_per_point];
+		double est[sets_per_point];
+
+		for (size_t i = 0; i < sets_per_point; ++i) {
+			struct hll *hll = hll_new(prec, HLL_DENSE);
+
+			for (size_t j = 0; j < card; ++j) {
+				uint64_t val = big_rand();
+				hll_add(hll, hash(val));
+			}
+			est[i] = hll_estimate(hll);
+			error[i] = fabs(est[i] - card);
+
+			hll_delete(hll);
+		}
+
+		double max_err = max_of(error, sets_per_point) / (card + 1);
+		max_err_sum += max_err;
+
+		double avg_est = average_sum_of(est, sets_per_point);
+		double std_err = dispersion_of(est, card, sets_per_point) /
+				(card + 1);
+		std_err_sum += std_err;
+
+		MAYBE_PRINT(output,
+			    "%2d, %12zu, %12.2f, %12lg, %12lg\n",
+			    prec, card, avg_est, std_err, max_err);
+	}
+
+	double avg_std_err = std_err_sum / n_points;
+	double avg_max_err = max_err_sum / n_points;
+
+	res->std_err = avg_std_err;
+	res->max_err = avg_max_err;
+	res->exp_err = hll_error(prec);
+}
+
+/*
+ * This test can dump the data that is used to measure
+ * the estimation error. These data can be used for further
+ * analysis and empirical based impovements of the algorithm.
+ */
+void
+test_dense_hyperloglog_error(void)
+{
+	header();
+	plan(HLL_MAX_PRECISION - HLL_MIN_PRECISION + 1);
+
+	struct est_errors errors[HLL_MAX_PRECISION + 1];
+
+	FILE *output = NULL;
+	if (OUTPUT_FILE_NAME != NULL) {
+		output = fopen(OUTPUT_FILE_NAME, "w");
+		assert(output && "Can't open the output file.");
+	}
+
+	MAYBE_PRINT(output, "prec, card, avg_est, std_err, max_err\n");
+
+	for (int prec = HLL_MIN_PRECISION;
+	     prec <= HLL_MAX_PRECISION; ++prec) {
+		size_t n_regs = 1u << prec;
+		size_t max_card = 10 * n_regs;
+		measure_dense_hll_estimation_error(prec, max_card,
+						   N_POINTS, SETS_PER_POINT,
+						   errors + prec, output);
+		/*
+		 * The error of HyperLogLog is close to 1/sqrt(n_counters),
+		 * but for small cardinalities LinearCounting is used because
+		 * it has better accuracy, so the resulting error must be
+		 * smaller than the HyperLogLog theoretical error.
+		 */
+		ok(errors[prec].std_err < errors[prec].exp_err,
+		   "The actual error doesn't exceed the expected value.");
+	}
+
+	if (output)
+		fflush(output);
+
+	for (int prec = HLL_MIN_PRECISION;
+	     prec <= HLL_MAX_PRECISION; ++prec) {
+		MAYBE_PRINT(output,
+			    "prec:%d, std_err:%lg, max_err:%lg, exp_err: %lg\n",
+			    prec, errors[prec].std_err, errors[prec].max_err,
+			    errors[prec].exp_err);
+	}
+
+	check_plan();
+	footer();
+}
+
+void
+test_sparse_hll_error(void)
+{
+	header();
+	plan(110);
```

Comment:
It was difficult to count the number of tests, so I just ran the test and got the number of tests performed.
I changed this test to avoid this magic number.